internal/cli/certificate.go

```diff
@@ -0,0 +1,367 @@
+package cli
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"net/http"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+	apiclient "github.com/flightctl/flightctl/internal/api/client"
+	"github.com/flightctl/flightctl/internal/client"
+	fccrypto "github.com/flightctl/flightctl/internal/crypto"
+	"github.com/google/uuid"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"golang.org/x/term"
+	"sigs.k8s.io/yaml"
+)
+
+const secondsInDay = 86400
+
+type CertificateOptions struct {
+	GlobalOptions
+	Name            string
+	SignerName      string
+	Expiration      string
+	OutputFormat    string
+	PrivKeyPassword string
+}
+
+func DefaultCertificateOptions() *CertificateOptions {
+	return &CertificateOptions{
+		GlobalOptions: DefaultGlobalOptions(),
+		SignerName:    "ca",
+		OutputFormat:  "components",
+	}
+}
+
+func NewCmdCertificate() *cobra.Command {
+	o := DefaultCertificateOptions()
+	cmd := &cobra.Command{
+		Use: "certificate [subcommand]",
+		// more subcommands may be added later
+		Short:     "request a new enrollment certificate for a device with 'certificate request'",
+		Args:      cobra.MatchAll(cobra.ExactArgs(1), cobra.OnlyValidArgs),
+		ValidArgs: []string{"request"},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Complete(cmd, args); err != nil {
+				return err
+			}
+			if err := o.Validate(args); err != nil {
+				return err
+			}
+			return o.Run(cmd.Context(), args)
+		},
+		SilenceUsage: true,
+	}
+	o.Bind(cmd.Flags())
+	return cmd
+}
+
+func (o *CertificateOptions) Bind(fs *pflag.FlagSet) {
+	o.GlobalOptions.Bind(fs)
+	fs.StringVarP(&o.Name, "name", "n", "mycsr", "specify a name for the certificate signing request")
+	fs.StringVarP(&o.SignerName, "signer", "s", "ca", "specify certificate signer - options currently supported: 'ca'")
+	fs.StringVarP(&o.Expiration, "expiration", "e", "7d", "specify desired certificate expiration in days, example: 7d")
+	fs.StringVarP(&o.OutputFormat, "output", "o", "components", "specify desired output format: either 'agent-config' to output values in one config file, or 'components' to output cert, key, and ca values in separate files")
+	fs.StringVarP(&o.PrivKeyPassword, "password", "w", "", "specify a password for newly generated private key, which will be encrypted on local disk")
```

Comment:
Passing in a password like this is insecure (it'll show in `ps` for example). Instead, we'd probably want to read it either from env var or from stdin (say, if the parameter was `-w -`).

Reply:
https://docs.openssl.org/1.0.2/man1/openssl/#encoding-and-cipher-commands this is what openssl accept, I think it's fine to accept cmdline, then it's up to the user to decide , also the fallback is "asking password", we could add env in addition.

Reply:
Yeah, I didn't love this either but wasn't sure whether to completely drop it as an option. I can add an env var option in addition to taking the pw at runtime or as a command line arg. I don't know if cobra will read from stdin? I can look into it...